File: Languages/go/low-level-design-lld/design-patterns/factory.go
package main

import (
	"fmt"
	"math"
)

// Factory Method: ShapeFactory creates shapes based on type string.
// New shapes can be added without modifying existing code (Open/Closed Principle).

// Shape is the product interface.
type Shape interface {
	Area() float64
	Perimeter() float64
	String() string
}

// Circle implementation
type Circle struct {
	Radius float64
}

func (c Circle) Area() float64      { return math.Pi * c.Radius * c.Radius }
func (c Circle) Perimeter() float64  { return 2 * math.Pi * c.Radius }
func (c Circle) String() string      { return fmt.Sprintf("Circle(radius=%.2f)", c.Radius) }

// Rectangle implementation
type Rectangle struct {
	Width, Height float64
}

func (r Rectangle) Area() float64      { return r.Width * r.Height }
func (r Rectangle) Perimeter() float64  { return 2 * (r.Width + r.Height) }
func (r Rectangle) String() string      { return fmt.Sprintf("Rectangle(%.2f x %.2f)", r.Width, r.Height) }

// ShapeFactory creates shapes by type.
type ShapeFactory struct{}

func (f ShapeFactory) Create(shapeType string, params ...float64) (Shape, error) {
	switch shapeType {
	case "circle":
		if len(params) < 1 {
			return nil, fmt.Errorf("circle requires radius")
		}
		return Circle{Radius: params[0]}, nil
	case "rectangle":
		if len(params) < 2 {
			return nil, fmt.Errorf("rectangle requires width and height")
		}
		return Rectangle{Width: params[0], Height: params[1]}, nil
	default:
		return nil, fmt.Errorf("unknown shape type: %s", shapeType)
	}
}

func main() {
	factory := ShapeFactory{}

	shapes := []struct {
		kind   string
		params []float64
	}{
		{"circle", []float64{5.0}},
		{"rectangle", []float64{3.0, 4.0}},
		{"circle", []float64{2.5}},
	}

	for _, s := range shapes {
		shape, err := factory.Create(s.kind, s.params...)
		if err != nil {
			fmt.Printf("Error: %v\n", err)
			continue
		}
		fmt.Printf("%s -> Area: %.2f, Perimeter: %.2f\n", shape, shape.Area(), shape.Perimeter())
	}
}

File: Languages/python/low-level-design-lld/design-patterns/factory.py
"""
Factory Method Pattern

Defines an interface for creating objects, letting subclasses decide which
class to instantiate. New shapes can be added without modifying ShapeFactory
by registering them at runtime.
"""

from abc import ABC, abstractmethod
import math


class Shape(ABC):
    @abstractmethod
    def area(self) -> float:
        pass

    @abstractmethod
    def describe(self) -> str:
        pass


class Circle(Shape):
    def __init__(self, radius: float):
        self.radius = radius

    def area(self) -> float:
        return math.pi * self.radius ** 2

    def describe(self) -> str:
        return f"Circle(radius={self.radius})"


class Rectangle(Shape):
    def __init__(self, width: float, height: float):
        self.width = width
        self.height = height

    def area(self) -> float:
        return self.width * self.height

    def describe(self) -> str:
        return f"Rectangle({self.width}x{self.height})"


class Triangle(Shape):
    def __init__(self, base: float, height: float):
        self.base = base
        self.height = height

    def area(self) -> float:
        return 0.5 * self.base * self.height

    def describe(self) -> str:
        return f"Triangle(base={self.base}, height={self.height})"


class ShapeFactory:
    """Open for extension: register new shapes without modifying factory code."""

    _registry: dict[str, type[Shape]] = {}

    @classmethod
    def register(cls, name: str, shape_cls: type[Shape]) -> None:
        cls._registry[name] = shape_cls

    @classmethod
    def create(cls, name: str, **kwargs) -> Shape:
        if name not in cls._registry:
            raise ValueError(f"Unknown shape: {name}")
        return cls._registry[name](**kwargs)


# Register built-in shapes
ShapeFactory.register("circle", Circle)
ShapeFactory.register("rectangle", Rectangle)
ShapeFactory.register("triangle", Triangle)


if __name__ == "__main__":
    circle = ShapeFactory.create("circle", radius=5)
    assert abs(circle.area() - 78.54) < 0.01
    assert circle.describe() == "Circle(radius=5)"

    rect = ShapeFactory.create("rectangle", width=3, height=4)
    assert rect.area() == 12.0

    tri = ShapeFactory.create("triangle", base=6, height=4)
    assert tri.area() == 12.0

    # Extend without modifying factory
    class Pentagon(Shape):
        def __init__(self, side: float):
            self.side = side

        def area(self) -> float:
            return (math.sqrt(5 * (5 + 2 * math.sqrt(5))) / 4) * self.side ** 2

        def describe(self) -> str:
            return f"Pentagon(side={self.side})"

    ShapeFactory.register("pentagon", Pentagon)
    pent = ShapeFactory.create("pentagon", side=3)
    assert pent.area() > 0

    print("All tests passed!")
